service/gcs/bridge/bridge.go

```diff
@@ -317,7 +340,55 @@ func (b *Bridge) PublishNotification(n *prot.ContainerNotification) {
 	b.responseChan <- resp
 }
 
+// negotiateProtocol was introduced in v4 so will not be called
+// with a minimum lower than that.
+func (b *Bridge) negotiateProtocol(w ResponseWriter, r *Request) {
+	if r.Version != prot.PvInvalid {
+		logrus.Errorf("bridge: Message Type: %v was called more than once.", r.Header.Type)
+		w.Error("", gcserr.NewHresultError(gcserr.HrVmcomputeProtocolError))
+		return
+	}
+
+	var request prot.NegotiateProtocol
+	if err := commonutils.UnmarshalJSONWithHresult(r.Message, &request); err != nil {
+		w.Error("", errors.Wrapf(err, "failed to unmarshal JSON in message \"%s\"", r.Message))
+		return
+	}
+
+	if request.MaximumVersion.Major < uint32(prot.PvV4) || uint32(prot.PvMax) < request.MinimumVersion.Major {
+		w.Error(request.ActivityID, gcserr.NewHresultError(gcserr.HrVmcomputeUnsupportedProtocolVersion))
+		return
+	}
+
+	min := func(x, y uint32) uint32 {
+		if x < y {
+			return x
+		}
+		return y
+	}
+
+	major := min(uint32(prot.PvMax), request.MaximumVersion.Major)
+
+	response := &prot.NegotiateProtocolResponse{
+		MessageResponseBase: &prot.MessageResponseBase{
+			ActivityID: request.ActivityID,
+		},
+		Version:      prot.Version{Major: major, Minor: 0, Patch: 0, Metadata: ""},
+		Capabilities: capabilities,
+	}
+
+	// Set our protocol selected version before return.
+	b.protVer = prot.ProtocolVersion(major)
```

Comment:
There's potentially a race condition here if two protocol negotiation/create messages are sent in quick succession. Maybe protect this with a mutex here and in `createContainer`?

Reply:
Spoke offline. Its ok to trust the HCS call chain here I think.